src/history/hs_cursor.c

```diff
@@ -18,56 +18,20 @@ __wt_hs_row_search(WT_CURSOR_BTREE *hs_cbt, WT_ITEM *srch_key, bool insert)
     WT_BTREE *hs_btree;
     WT_DECL_RET;
     WT_SESSION_IMPL *session;
-    bool leaf_found;
-#ifdef HAVE_DIAGNOSTIC
-    WT_PAGE *page;
-#endif
 
     hs_btree = CUR2BT(hs_cbt);
     session = CUR2S(hs_cbt);
-    leaf_found = false;
 
     /*
      * Check whether the search key can be find in the provided leaf page, if exists. Otherwise
      * perform a full search.
      */
-    if (__wt_cursor_page_pinned(hs_cbt, true)) {
-#ifdef HAVE_DIAGNOSTIC
-        WT_ORDERED_READ(page, hs_cbt->ref->page);
-#endif
-        /*
-         * The page must be pinned and we should have a hazard pointer on that. Ensure the page is
-         * not evictable.
-         */
-        WT_ASSERT(session, __wt_hazard_check(session, hs_cbt->ref, NULL) != NULL);
-        WT_WITH_BTREE(session, hs_btree,
-          ret = __wt_row_search(hs_cbt, srch_key, insert, hs_cbt->ref, false, &leaf_found));
-        WT_ERR(ret);
-
-        /*
-         * Only use the pinned page search results if search returns an exact match or a slot other
-         * than the page's boundary slots, if that's not the case, the record might belong on an
-         * entirely different page.
-         */
-        if (leaf_found &&
-          (hs_cbt->compare != 0 &&
-            (hs_cbt->slot == 0 || hs_cbt->slot == hs_cbt->ref->page->entries - 1)))
-            leaf_found = false;
-
-        /* Ensure there is no eviction happened on this page. */
-        WT_ASSERT(session, page == hs_cbt->ref->page);
-    }
+    if (__wt_cursor_page_pinned(hs_cbt, false))
+        return (0);
```

Comment:
With the change, I think the retry part is no longer needed. This was required before to point the return value back to the newly updated value because in the old code, the search will overwrite the return value.